packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -222,10 +225,66 @@ class UpgradeCommandRunner {
     }
   }
 
+  /// Checks if the Flutter git repository is tracking a "standard remote".
+  ///
+  /// Using `flutter upgrade` is not supported from a non-standard remote. A git
+  /// remote should have the same url as [_flutterGit] to be considered as a
+  /// "standard" remote.
+  ///
+  /// Exits tool if the tracking remote is not standard.
+  void verifyStandardRemote(FlutterVersion localVersion) {
+    // Strip `.git` suffix from repository url and _flutterGit
+    final String trackingUrl = _stripDotGit(localVersion.repositoryUrl);
+    final String flutterGitUrl = _stripDotGit(_flutterGit);
+
+    if (trackingUrl != flutterGitUrl) {
+      if (globals.platform.environment.containsKey('FLUTTER_GIT_URL')) {
+        // If `FLUTTER_GIT_URL` is set, inform the user to either remove the
+        // `FLUTTER_GIT_URL` environment variable or set it to the current
+        // tracking remote to continue.
+        throwToolExit(
+          'Unable to upgrade Flutter: The Flutter SDK is tracking '
+          '"${localVersion.repositoryUrl}" but "FLUTTER_GIT_URL" is set to "$_flutterGit".\n'
+          'Either remove "FLUTTER_GIT_URL" from the environment or set "FLUTTER_GIT_URL" '
+          'to "${localVersion.repositoryUrl}", and retry.\n\n'
+          'As an alternative, if you are okay with losing local changes you made '
+          'to the SDK, re-install Flutter by going to https://flutter.dev/docs/get-started/install.'
+        );
+      } else {
+        // Inform that the user has to either set the environment variable, or
+        // change the url of the tracking remote to the Flutter GitHub repository
+        // to continue.
+        throwToolExit(
+          'Unable to upgrade Flutter: The Flutter SDK is tracking a non-standard '
+          'remote "${localVersion.repositoryUrl}".\n'
+          '  - To use the current remote, set the environment variable "FLUTTER_GIT_URL" '
```

Comment:
Even though this message is correct in offering all the options, I think it will be confusing, especially for new users who are not familiar with git. I think the first suggestion we make should be going to the website and re-installing flutter, to give new users the best experience. As a fallback we should say:

`"If you are intentionally tracking a non-standard git remote, you should keep your repository up to date directly with git."`

Reply:
In general, if they are maintaining a non-standard git setup, it is their responsibility to manage their repo with git themselves. It's impossible for us to anticipate and provide the correct git commands for every possible setup.

Reply:
I tried to create a table listing some of the probable setups [here](https://github.com/flutter/flutter/pull/79372#discussion_r605588449).

Reply:
I removed the git commands and just mentioned "git remote" instead to change the upstream url.
Ditto for the comments below.